core/base/arrayPreconditioning/ArrayPreconditioning.h

```diff
@@ -0,0 +1,593 @@
+/// \ingroup base
+/// \class ttk::ArrayPreconditioning
+/// \author Michael Will <mswill@rhrk.uni-kl.de>
+/// \date 2022.
+///
+/// This module defines the %ArrayPreconditioning class that generates order
+/// arrays from a selection of scalar field arrays.
+///
+
+#pragma once
+
+// ttk common includes
+#include <Debug.h>
+
+#ifdef TTK_ENABLE_MPI
+#include <mpi.h>
+#endif
+
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+namespace ttk {
+
+  /**
+   * The ArrayPreconditioning class provides methods to generate order arrays
+   * from a selection of scalar field arrays.
+   */
+  class ArrayPreconditioning : virtual public Debug {
+
+  public:
+    ArrayPreconditioning();
+#ifdef TTK_ENABLE_MPI
+    MPI_Datatype getMPIType(const float ttkNotUsed(val)) const {
+      return MPI_FLOAT;
+    };
+    MPI_Datatype getMPIType(const int ttkNotUsed(val)) const {
+      return MPI_INT;
+    };
+    MPI_Datatype getMPIType(const unsigned int ttkNotUsed(val)) const {
+      return MPI_UNSIGNED;
+    };
+    MPI_Datatype getMPIType(const double ttkNotUsed(val)) const {
+      return MPI_DOUBLE;
+    };
+    MPI_Datatype getMPIType(const long double ttkNotUsed(val)) const {
+      return MPI_LONG_DOUBLE;
+    };
+    MPI_Datatype getMPIType(const long ttkNotUsed(val)) const {
+      return MPI_LONG;
+    };
+    MPI_Datatype getMPIType(const unsigned long ttkNotUsed(val)) const {
+      return MPI_UNSIGNED_LONG;
+    };
+    MPI_Datatype getMPIType(const long long ttkNotUsed(val)) const {
+      return MPI_LONG_LONG;
+    };
+    MPI_Datatype getMPIType(const unsigned long long ttkNotUsed(val)) const {
+      return MPI_UNSIGNED_LONG_LONG;
+    };
+#endif
+    template <typename DT, typename IT>
+    struct value {
+      DT scalar;
+      IT globalId;
+      IT localId;
+      IT ordering = -1;
+
+      value(DT _scalar, IT _globalId, IT _localId)
+        : scalar(_scalar), globalId(_globalId), localId(_localId) {
+      }
+    };
+
+    // creates a vector of value structs based on the given pointers
+    // only takes values of vertices which mainly belong to the current rank
+    // ( so only vertices which are no ghost cells)
+    template <typename DT, typename IT>
+    void populateVector(std::vector<value<DT, IT>> &valuesToSortVector,
+                        std::vector<IT> &gidsToGetVector,
+                        std::unordered_map<IT, IT> &gidToLidMap,
```

Comment:
NOTE for later: this variable (`gidToLidMap`) should be a member of all the classes deriving from the `BaseClass` (if MPI is detected and enabled).

Reply:
NOTE for late (@eve-le-guillou): it looks like this should be embedded in the triangulation class.